src/c/c_dataset.cpp

```diff
@@ -30,163 +30,196 @@
-                     CMetaDataType type)
+void add_meta_scalar(void *dataset,
+                     const char *name,
+                     const size_t name_length,
+                     const void *data,
+                     const CMetaDataType type)
 {
-  /* Add a meta data value to the named meta data field
-  */
-  DataSet* d = (DataSet*)dataset;
-  std::string name_str = std::string(name, name_length);
+  // Sanity check parameters
+  if (NULL == dataset || NULL == name || NULL == data)
+    return;
+
+  DataSet *d = reinterpret_cast<DataSet *>(dataset);
+  std::string name_str(name, name_length);
+
   d->add_meta_scalar(name_str, data,
                      convert_metadata_type(type));
-  return;
 }
 
+// Add a meta data value to the named meta data field
 extern "C"
-void add_meta_string(void* dataset,
-                     const char* name,
-                     size_t name_length,
-                     const char* data,
+void add_meta_string(void *dataset,
+                     const char *name,
+                     const size_t name_length,
+                     const char *data,
                      const size_t data_length)
 {
-  /* Add a meta data value to the named meta data field
-  */
-  DataSet* d = (DataSet*)dataset;
-  std::string name_str = std::string(name, name_length);
-  std::string data_str = std::string(data, data_length);
-  d->add_meta_string(name_str, data_str);
+  // Sanity check parameters
+  if (NULL == dataset || NULL == name || NULL == data)
+    return;
 
-  return;
+  DataSet *d = reinterpret_cast<DataSet *>(dataset);
+  std::string name_str(name, name_length);
+  std::string data_str(data, data_length);
+
+  d->add_meta_string(name_str, data_str);
 }
 
+// Get a tensor of a specified type from the database.
+// This function may allocate new memory for the tensor.
+// This memory will be deleted when the user deletes the
+// DataSet object.
 extern "C"
-void get_dataset_tensor(void* dataset,
-                        const char* name,
+void get_dataset_tensor(void *dataset,
+                        const char *name,
                         const size_t name_length,
-                        void** data,
-                        size_t** dims,
-                        size_t* n_dims,
-                        CTensorType* type,
-                        CMemoryLayout mem_layout)
+                        void* *data,
+                        size_t* *dims,
+                        size_t *n_dims,
+                        CTensorType *type,
+                        const CMemoryLayout mem_layout)
 {
-  /* Get a tensor of a specified type from the database.
-  This function may allocate new memory for the tensor.
-  This memory will be deleted when the user deletes the
-  DataSet object.
-  */
-  DataSet* d = (DataSet*)dataset;
-  std::string name_str = std::string(name, name_length);
-
-  TensorType t_type;
+  // Sanity check parameters
+  if (NULL == dataset || NULL == name || NULL == data ||
```

Comment:
Can't ``data`` be NULL since we are allocating the memory for them?  It seems totally fine for user to pass in ``NULL``.